src/main/java/io/bf2/kafka/authorizer/CustomAclAuthorizer.java

```diff
@@ -246,6 +247,35 @@ void addAllowedListeners(Map<String, ?> configs) {
         }
     }
 
+    /**
+     * To avoid misleading users we delete principals that are explicitly targeted by CustomAclAuthorizer rules.
+     * It is possible for a user with ACL rules in Zookeeper to become a user managed by CustomAclAuthorizer, in which
+     * case we would never use or update rules for them in ZK. The old rulesremain, dangling, in ZK and visible through
+     * kafka ACL tooling. This would mislead users becuase there will be a mismatch between what the ZK rules say they
+     * should be permitted to do and the truth.
+     */
+    private void deleteOrphanBindingsFromDelegate() {
+        List<AclBindingFilter> toDelete = aclPrincipals.stream().map(aclPrincipal -> {
+            AccessControlEntryFilter principalFilter = new AccessControlEntryFilter(aclPrincipal, null, AclOperation.ANY, AclPermissionType.ANY);
+            return new AclBindingFilter(ResourcePatternFilter.ANY, principalFilter);
+        }).collect(Collectors.toList());
+        if (!toDelete.isEmpty()) {
+            log.info("deleting orphaned bindings if present, using filter: {}", toDelete);
+            List<CompletableFuture<AclDeleteResult>> deletionResults = delegate.deleteAcls(null, toDelete)
+                    .stream().map(CompletionStage::toCompletableFuture).collect(Collectors.toList());
+            CompletableFuture<Void> future = CompletableFuture.allOf(deletionResults.toArray(CompletableFuture[]::new));
+            future.whenComplete((unused, throwable) -> {
+                if (throwable == null) {
+                    log.info("deleting orphaned bindings completed successfully");
```

Comment:
could log out the `deletionResults` to see what bindings were deleted

Reply:
That's a good point, I think we should be logging the deletes. I'm not sure if it is should be in the standard log or if it should be done via the AuditLog (aka [`io.bf2.kafka.authorizer.AuditLoggingController#logAuditMessage`](https://github.com/bf2fc6cc711aee1a0c2a/kas-broker-plugins/blob/main/src/main/java/io/bf2/kafka/authorizer/AuditLoggingController.java#L110)). 



@k-wall do you have any thoughts about if and how the ownership rule changes should manifest in the Audit log? Arguably it is a user generated CRUD event but, at this point in the process we have lost any hook into the actual user performing the action.

Reply:
Presumably the fleet manager audits the change in ownership, @pb82 would be able to tell us for sure.

>> could log out the deletionResults to see what bindings were deleted
> That's a good point, I think we should be logging the deletes. I'm not sure if it is should be in the standard log or if it should be done via the AuditLog (aka [io.bf2.kafka.authorizer.AuditLoggingController#logAuditMessage](https://github.com/bf2fc6cc711aee1a0c2a/kas-broker-plugins/blob/main/src/main/java/io/bf2/kafka/authorizer/AuditLoggingController.java#L110)).

I think logging the fact that the clean-up has happened is a good idea.  I think I'd just use a custom info level message would be sufficient.

Reply:
Have added standard logging at info for successes and error for any exceptionally completed futures/results. A bit of a gnarly chain of `future -> delete result -> List[delete binding result]` so the logging code ended up being quite large. 

If this got any bigger we would want to extract an `OrphanCleaner` class out.

Reply:
looks reasonable to me.